projects/utils-maps/src/lib/ol/customRenderers/interpolation_renderer.ts

```diff
@@ -0,0 +1,600 @@
+    }
+
+    updateParas(power: number, smooth: boolean, showLabels: boolean): void {
+        const newSettings: InterpolationRendererSettings = {
+            colorRamp: this.options.renderSettings.colorRamp,
+            maxEdgeLength: this.options.renderSettings.maxEdgeLength,
+            storeInterpolatedPixelData: this.options.renderSettings.storeInterpolatedPixelData,
+            valueProperty: this.options.renderSettings.valueProperty,
+            power: power,
+            showLabels: showLabels,
+            smooth: smooth,
+        };
+        (super.getRenderer() as InterpolationRenderer).updateSettings(newSettings);
+        this.options.renderSettings = newSettings;
+    }
+}
+
+/**
+ * This renderer runs three shaders in a row.
+ *  1. interpolationShader: takes every observation at every pixel and executes the interpolation. The values are stored in `valueFb`.
+ *  2. colorizationShader: uses the interpolated values from valueFb to apply the colorization according to the given colorRamp and smoothing-options.
+ *  3. arrangementShader: the previous shaders have moved the data in the center of the canvas. this shader now arranges the pixels to the correct position relative to the map.
+ *
+ * Only the third shader needs to be executed with every frame. This way, the operation-heavy interpolation does not slow down the map.
+ * It generally makes sense to arrange shaders in such a way that all openlayers-perspective-operations occur in the last shader.
+ *
+ * valueFb is also being used to handle click events: from this structure we get the actual value at a pixel when the user clicks.
+ *
+ * Note a few caveats.
+ * This implementation is not really intended for updating observations, maxEdgeLength or colorRamps at runtime. These parameters are rather intended for the developer to set once.
+ * While you can change the color-ramp at runtime, it's length is hardcoded in the colorization shader, so you'd have to recompile it to properly reflect the new ramp.
+ * In the same way, the interpolation-shader has the number of observations baked into it. When new data becomes available, you must recompile the interpolation shader.
+ */
+export class InterpolationRenderer extends LayerRenderer<VectorLayer> {
+
+    private container: HTMLDivElement;
+    private pointRenderer: CanvasVectorLayerRenderer;
+    private webGlCanvas: HTMLCanvasElement;
+    private gl: WebGLRenderingContext;
+    private interpolationShader: Shader;
+    private valueFb: Framebuffer;
+    private colorizationShader: Shader;
+    private colorFb: Framebuffer;
+    private arrangementShader: Shader;
+    private interpolatedValues: Uint8Array;
+
+    private projection: Projection;
+    private bbox: number[];
+
+    constructor(layer: VectorLayer, private settings: InterpolationRendererSettings) {
+        super(layer);
+
+        // setting up HTML element
+        this.container = document.createElement('div');
```

Comment:
When interacting with the browser DOM the `angular document`  https://angular.io/api/common/DOCUMENT should be used.
```ts
import { Inject } from "@angular/core";
import { DOCUMENT } from '@angular/common'
...
constructor(@Inject(DOCUMENT) private document: Document)
...
```